Fix Test Flakiness by adding short sleep in TestMetricsRefresh

LukeAVanDrie opened the pull request:
This PR addresses flakiness observed in the TestMetricsRefresh test within the pkg/epp/backend/metrics package.

### Problem

The root cause of the flakiness was a race condition stemming from the asynchronous nature of the `StopRefreshLoop` method. The test would:
1. Signal the metrics refresh goroutine to stop via `StopRefreshLoop()`.
2. Immediately update the mock `PodMetricsClient` to return different metric values.
3. Assert that the `PodMetrics` instance still held the original metrics.

Occasionally, the refresh goroutine would execute one final time after `StopRefreshLoop()` was called but before it fully terminated, picking up the new metrics. This led to the assertion failing.

### Solution

To resolve this, I added a `time.Sleep(pmf.refreshMetricsInterval * 2 /* small buffer for robustness */)` before any test assertions are made. This causes the test to sleep for just 2ms. I also added a `stopOnce` around closing the `done` channel for robustness.

kfswain commented:
Fixes: https://github.com/kubernetes-sigs/gateway-api-inference-extension/issues/719

kfswain commented:
/ok-to-test

nirrozenbaum commented:
I agree with @liu-cong's general point, making this function synchronous doesn't sound to me like the right solution.
thinking forward about data layer and extensible scrapers, we might need to run multiple routines for different scrapers.
I wouldn't touch the existing pod metrics code. this is an issue with the test itself. 
@LukeAVanDrie your analysis of the issue is great:
> 1. Signal the metrics refresh goroutine to stop via StopRefreshLoop().
> 2. Immediately update the mock PodMetricsClient to return different metric values.
> 3. Assert that the PodMetrics instance still held the original metrics.
> Occasionally, the refresh goroutine would execute one final time after StopRefreshLoop() was called but before it fully terminated, picking up the new metrics. This led to the assertion failing.

I think a very easy and straight forward fix to the above problem is to let the test wait until the final call is completed.
taking the scraping timeout as a worst case is sufficient.  
so I would fix the bug by adding this single code line in the test file:
```go
pm.StopRefreshLoop()
time.Sleep(fetchMetricsTimeout) // THIS IS THE ONLY ADDED LINE
pmc.SetRes(map[types.NamespacedName]*MetricsState{namespacedName: updated})
```

liu-cong commented:
Nir's suggestion makes sense. Though in theory we could still run into a
flakiness due to the async nature, adding that timeout should reduce
flakiness to near zero in practice I believe.

On Tue, May 13, 2025 at 12:34 AM Nir Rozenbaum ***@***.***>
wrote:

> *nirrozenbaum* left a comment
> (kubernetes-sigs/gateway-api-inference-extension#824)
> <https://github.com/kubernetes-sigs/gateway-api-inference-extension/pull/824#issuecomment-2875377324>
>
> I agree with @liu-cong <https://github.com/liu-cong>'s general point,
> making this function synchronous doesn't sound to me like the right
> solution.
> thinking forward about data layer and extensible scrapers, we might need
> to run multiple routines for different scrapers.
> I wouldn't touch the existing pod metrics code. this is an issue with the
> test itself.
> @LukeAVanDrie <https://github.com/LukeAVanDrie> your analysis of the
> issue is great:
>
>
>    1. Signal the metrics refresh goroutine to stop via StopRefreshLoop().
>    2. Immediately update the mock PodMetricsClient to return different
>    metric values.
>    3. Assert that the PodMetrics instance still held the original metrics.
>    Occasionally, the refresh goroutine would execute one final time after
>    StopRefreshLoop() was called but before it fully terminated, picking up the
>    new metrics. This led to the assertion failing.
>
> I think a very easy and straight forward fix to the above problem is to
> let the test wait until the final call is completed.
> taking the scraping timeout as a worst case is sufficient.
> so I would fix the bug by adding this single code line in the test file:
>
> pm.StopRefreshLoop()time.Sleep(fetchMetricsTimeout) // THIS IS THE ONLY ADDED LINEpmc.SetRes(map[types.NamespacedName]*MetricsState{namespacedName: updated})
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes-sigs/gateway-api-inference-extension/pull/824#issuecomment-2875377324>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABUVECT6YT2YXAXQKOPUYEL26GOA5AVCNFSM6AAAAAB467IC3WVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQNZVGM3TOMZSGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> <kubernetes-sigs/gateway-api-inference-extension/pull/824/c2875377324@
> github.com>
>

kfswain commented:
> time.Sleep(fetchMetricsTimeout) // THIS IS THE ONLY ADDED LINE

this is what i was going to suggest. This solves the problem of the test flake and if we want to change how the metrics loop is handled we can do that later

LukeAVanDrie commented:
> I think a very easy and straight forward fix to the above problem is to let the test wait until the final call is completed. taking the scraping timeout as a worst case is sufficient. so I would fix the flakiness by adding this single code line in the test file:
> 
> ```go
> pm.StopRefreshLoop()
> time.Sleep(fetchMetricsTimeout) // THIS IS THE ONLY ADDED LINE
> pmc.SetRes(map[types.NamespacedName]*MetricsState{namespacedName: updated})
> ```

I incorporated this change and removed the wait group. I did, however, leave the `stopOnce` for robustness, but I can also remove this if you feel it is unnecessary.

LukeAVanDrie commented on pkg/epp/backend/metrics/pod_metrics.go:
```diff
@@ -138,5 +142,8 @@ func (pm *podMetrics) refreshMetrics() error {
+	pm.stopOnce.Do(func() {
+		close(pm.done)
+	})
+	pm.wg.Wait()
```
Ack, incorporated this feedback. Thanks for calling this out.

liu-cong commented:
/lgtm

kfswain commented:
/approve

Thanks!

LukeAVanDrie commented:
/fixes #719

kfswain commented:
Ah, yeah no problem! Most of our slash commands are a signal for the `k8s-ci-robot` to take some sort of action. the fixes or resolves keyword is a GH concept so no slash: https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests#linking-a-pull-request-to-an-issue

Either way, I manually closed the issue, thanks for doin this
